driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/vault/ClientEncryptionImpl.java

```diff
@@ -183,6 +198,85 @@ public Publisher<RewrapManyDataKeyResult> rewrapManyDataKey(final Bson filter, f
+                        // here we rely on the `createDataKey` publisher being cold, i.e., doing nothing until it is subscribed to
+                        .map(field -> Mono.fromDirect(createDataKey(kmsProvider, dataKeyOptions))
+                                // This is the closest we can do with reactive streams to setting the `dataKeyMayBeCreated` flag
+                                // immediately before calling `createDataKey`.
+                                .doOnSubscribe(subscription -> dataKeyMayBeCreated.set(true))
+                                .map(dataKey -> new SimpleImmutableEntry<>(field, dataKey))
+                        )
+                        .iterator();
+                Flux<SimpleImmutableEntry<BsonDocument, BsonBinary>> publisherOfFieldsAndDataKeyIds = Flux.concat(publishersOfFieldsAndDataKeyIds)
+                        .doOnNext(fieldAndDataKeyId -> {
+                            BsonDocument field = fieldAndDataKeyId.getKey();
+                            BsonBinary dataKeyId = fieldAndDataKeyId.getValue();
+                            field.put(keyIdBsonKey, dataKeyId);
+                        });
+                createCollectionMono = publisherOfFieldsAndDataKeyIds
+                        // All write actions in `doOnNext` above happen-before the completion (`onComplete`/`onError`) signals
+                        // for this publisher, because all signals are serial. `thenEmpty` further guarantees that the completion signal
+                        // for this publisher happens-before the `onSubscribe` signal for the publisher passed to it (the next publisher).
+                        // `defer` defers calling `createCollection` until the next publisher is subscribed to.
+                        // Therefore, all write actions in `doOnNext` above happen-before the invocation of `createCollection`,
+                        // which means `createCollection` is guaranteed to observe all those write actions, i.e.,
+                        // it is guaranteed to observe the updated document via the `actualEncryptedFields` reference.
+                        // Similarly, the updated document is observed in all other places below via the `actualEncryptedFields` reference.
+                        .thenEmpty(Mono.defer(() -> Mono.fromDirect(database.createCollection(collectionName,
+                                createCollectionOptions.clone().encryptedFields(actualEncryptedFields))))
+                        )
+                        .doOnError(RuntimeException.class, e -> {
+                            if (dataKeyMayBeCreated.get()) {
+                                throw new MongoUpdatedEncryptedFieldsException(actualEncryptedFields, format("Failed to create %s.", namespace), e);
+                            } else {
+                                throw e;
+                            }
+                        });
```

Comment:
Exceptions from `doOnError`/`defer` are handled by the Project Reactor such that the reactive streams protocol is not violated. I am not sure what is a better / more idiomatic way of transforming the error.

P.S. While writing this comment, I tried to find a better way again and found `Flux.handle`. Is this what is supposed to be used?

Reply:
Note that `doOnError` suppresses the original `e` when we throw `MongoUpdatedEncryptedFieldsException`, instead of treating the thrown exception as a replacement for the original one. As a result, we have `e` as both the cause of `MongoUpdatedEncryptedFieldsException` and an element in the suppressed set, which causes ugly stack-traces and implies that we are abusing `doOnError`. However, it may be that the Reactor simply does not support what we need.

Reply:
@rozza Thank you for pointing out the `onErrorMap` method, the emitted exceptions are nice now. There is no excuse for how I failed to find it myself.